### DISCUSSION TITLE
What is the correct way to run shell scripts in a hook in v9?

### DISCUSSION THREAD

**Discussion by janeklb** [typicode/husky#1543] [Q&A]
In v8 etc we could put anything in the hook file

eg.
```shell
#!/bin/sh
. "$(dirname "$0")/_/husky.sh"

echo "Doing step 1 of pre-commit hook"
npx ....


echo "Doing step 2 of pre-commit hook"
npx ...
```

What's the correct way to achieve this in v9?

If i remove the shebang header and `. "$(dirname "$0") ...`, then the hooks fails because it doesn't understand `echo "..."`. If I put the shebang back in then seems to work ok..

```shell
#!/bin/sh

echo "Doing step 1 of pre-commit hook"
npx ....


echo "Doing step 2 of pre-commit hook"
npx ...
```

However, if I remove the `_` directory (rm -rf `.husky/_`) and run `npm run prepare` (initialising husky), then the `_` is re-created. Should I be sourcing this directory in the hook script (as before), or is it purely optional now?

**Comment by ignaciobockl** [comment]
Hi @janeklb

In Husky v9, hooks are plain POSIX shell scripts — no shebang or sourcing needed.

### What changed in v9

The `.husky/_` directory is a Husky internal directory. You don't need to source it. Hook scripts are simple shell commands:

```bash
echo "Step 1"
npx lint-staged

echo "Step 2"
npx ...
```

### No shebang required

In v9, Husky executes hooks directly. The shebang and the sourcing line from v8 are no longer needed.

### If you need Bash features

If your hook requires Bash-specific syntax, use this pattern:

```bash
bash << EOF
# Your bash script here
if [ "$SOME_VAR" = "value" ]; then
  echo "Conditional step"
fi
EOF
```

### Testing

After editing, test with:

```bash
git commit --allow-empty -m "test hook"
```

The `exit 1` trick still works to test without committing.
